data/src/main/java/me/androidbox/data/worker_manager/SyncFullAgendaItemsWorker.kt

```diff
@@ -0,0 +1,127 @@
+
+        val responseResult = checkResult {
+            agendaService.fullAgenda()
+        }
+
+        if(runAttemptCount > RETRY_COUNT) {
+            Result.failure()
+        }
+
+        val result = responseResult.fold(
+            onSuccess = { fullAgendaDto ->
+                insertAllAgendaItems(fullAgendaDto, eventDao, taskDao, reminderDao)
+                Result.success()
+            },
+            onFailure = {
+                Result.retry()
+            }
+        )
+
+        return result
+    }
+}
+
+private suspend fun insertAllAgendaItems(fullAgendaDto: FullAgendaDto, eventDao: EventDao, taskDao: TaskDao, reminderDao: ReminderDao) {
+    supervisorScope {
+        val eventJob = launch {
+            fullAgendaDto.events.mapAsync { eventDto ->
+                val eventEntity = EventEntity(
+                    id = eventDto.id,
+                    title = eventDto.title,
+                    description = eventDto.description,
+                    startDateTime = eventDto.from,
+                    endDateTime = eventDto.to,
+                    remindAt = eventDto.remindAt,
+                    eventCreatorId = eventDto.host,
+                    isUserEventCreator = eventDto.isUserEventCreator,
+                    photos = eventDto.photos.map { photoDto ->  photoDto.key },
+                    isGoing = true,
+                    attendees = eventDto.attendees.map { attendeeDto ->
+                        attendeeDto.toAttendee()
+                    }
+                )
+
+                eventDao.insertEvent(eventEntity)
+            }
+        }
+
+        val taskJob = launch {
+            fullAgendaDto.tasks.mapAsync { taskDto ->
+                val taskEntity = TaskEntity(
+                    id = taskDto.id,
+                    title = taskDto.title,
+                    description = taskDto.description,
+                    time = taskDto.time,
+                    remindAt = taskDto.remindAt,
+                    isDone = taskDto.isDone
+                )
+
+                taskDao.insertTask(taskEntity)
+            }
+        }
+
+        val reminderJob = launch {
+            fullAgendaDto.reminders.mapAsync { reminderDto ->
+                val reminderEntity = ReminderEntity(
+                    id = reminderDto.id,
+                    title = reminderDto.title,
+                    description = reminderDto.description,
+                    time = reminderDto.time,
+                    remindAt = reminderDto.remindAt
+                )
+
+                reminderDao.insertReminder(reminderEntity)
+            }
+        }
+
+        eventJob.join()
+        taskJob.join()
+        reminderJob.join()
+    }
+}
+
+suspend fun <T, R> List<T>.mapAsync(transformation: suspend (T) -> R): List<R> {
+    return coroutineScope {
+        this@mapAsync.map { item ->
+            async {
+                transformation(item)
+            }
+        }.awaitAll()
+    }
+}
```

Comment:
I got this method from the following website. I thought I would try it out. 
https://kt.academy/article/cc-recipes

Reply:
No need for this tbh. Since you are already running each stuff in it's own job, inside a supervisorScope
Ex:
```
val eventJob = launch {
```

You are creating a new supervisorScope inside the job
It's really not necessary

If you apply the changes I mentioned above, you can actually remove the mapAsync

Reply:
You can have such a helper function, but it limits a bit the flexibility since it forces you to use `coroutineScope` and to await directly afterwards. I'd rather return a list of deferreds. If you need to map multiple lists you usually want to call await on all of them after mapping all of them.